### PR TITLE
Implementation of the SecurityService #LMR-203

### PR DESCRIPTION
**Pull request by martinglova**
Tha last pulls had to do something with URLs for my tests. I think there could be conflicts.

I've defined URLs as in the updated comment: https://lumeer.myjetbrains.com/youtrack/issue/LMR-203#comment=100-114

But I temporarily added `/roles` before all of the URLs in the code to pass all the tests.

### CODE REVIEW

File: api/src/main/java/io/lumeer/engine/api/dto/Roles.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.DataFormatException;
+/**
+ * @author <a href="mailto:martin.glova@lumeer.io">Martin Glova</a>
+ */
+public class Roles extends DataDocument {
+   public Roles(final DataDocument data, final String resource) throws DataFormatException {
```
Comment (kubedo8) [inline api/src/main/java/io/lumeer/engine/api/dto/Roles.java]: DTO objects should look like others. So it have to contains string,integer and other needed fields..

File: war/src/main/java/io/lumeer/engine/controller/OrganizationFacade.java [hunk 2]
```diff
@@ -127,7 +128,7 @@ public Organization readOrganization(final String organizationCode) {
     *       organization data
     * @return id of the organization
     */
-   public String createOrganization(final Organization organization) {
+   public String createOrganization(final Organization organization) throws InvalidDocumentKeyException {
       String id = dataStorage.createDocument(LumeerConst.Organization.COLLECTION_NAME, organization.toDataDocument());
       databaseInitializer.onOrganizationCreated(id);
       securityFacade.addOrganizationUserRole(organization.getCode(), userFacade.getUserEmail(), LumeerConst.Security.ROLE_MANAGE);
```
Comment (kubedo8) [inline war/src/main/java/io/lumeer/engine/controller/OrganizationFacade.java]: Move this line to `databaseInitilizer.onOrganizationCreated` method and don't throw exception because we know, that we add correct role.

File: war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java [hunk 3]
```diff
@@ -147,7 +148,7 @@ public Project readProject(final String projectCode) {
     *       project data
     * @return id of the organization
     */
-   public String createProject(final Project project) {
+   public String createProject(final Project project) throws InvalidDocumentKeyException {
       DataDocument dataDocument = project.toDataDocument()
                                          .append(LumeerConst.Project.ATTR_ORGANIZATION_ID, organizationFacade.getOrganizationId());
       String id = dataStorage.createDocument(LumeerConst.Project.COLLECTION_NAME, dataDocument);
```
Comment (kubedo8) [inline war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java]: Same here. Move adding role to `databaseInitilizer.onOrganizationCreated` method and don't throw exception because we know, that we add correct role.

File: war/src/main/java/io/lumeer/engine/controller/ViewFacade.java [hunk 4]
```diff
@@ -83,7 +84,7 @@ public int createView(
          final String originalViewName,
          final String viewType,
          final String description,
-         final DataDocument configuration) throws ViewAlreadyExistsException {
+         final DataDocument configuration) throws ViewAlreadyExistsException, InvalidDocumentKeyException {
```
Comment (kubedo8) [inline war/src/main/java/io/lumeer/engine/controller/ViewFacade.java]: Same here...

File: war/src/main/java/io/lumeer/engine/controller/OrganizationFacade.java (hunk 2 above)
Comment (martinglova) [inline war/src/main/java/io/lumeer/engine/controller/OrganizationFacade.java]: What do you mean by move this line? It is a line with the declaration of a method.

But finally, I see that data are not dependent on input so I suggest to do:

`try {`
`   securityFacade.addOrganizationUserRole(organization.getCode(), userFacade.getUserEmail(), LumeerConst.Security.ROLE_MANAGE);`
`} catch (InvalidDocumentKeyException e) {`
`   // Never can happen`
`}`

File: war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java (hunk 3 above)
Comment (martinglova) [inline war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java]: Same here...

File: war/src/main/java/io/lumeer/engine/controller/ViewFacade.java (hunk 4 above)
Comment (martinglova) [inline war/src/main/java/io/lumeer/engine/controller/ViewFacade.java]: Same here...

File: war/src/main/java/io/lumeer/engine/controller/SecurityFacade.java [hunk 5]
```diff
@@ -342,4 +345,53 @@ private String userOrGroupKey(String user) {
...
+      return getRolesList(doc);
+   }
+
+   private List<Role> getRolesList(DataDocument doc) {
+      List<Role> output = new ArrayList<>();
+      doc.getDataDocument(LumeerConst.Security.ROLES_KEY).forEach((k, v)-> {
+         output.add(new Role(
+               k,
+               ((Map<String, List<String>>) v).get(LumeerConst.Security.USERS_KEY),
+               ((Map<String, List<String>>) v).get(LumeerConst.Security.GROUP_KEY)));
```
Comment (livthomas) [inline war/src/main/java/io/lumeer/engine/controller/SecurityFacade.java]: `Role` class should have a constructor that takes `DataDocument` as a parameter in order to make this method simpler.

File: war/src/test/java/io/lumeer/engine/controller/SecurityFacadeIntegrationTest.java [hunk 6]
```diff
@@ -106,8 +109,24 @@ public void addAndRemoveOrganizationUserRole() throws Exception {
       securityFacade.addOrganizationUserRole(org, user, roleManage);
       assertThat(securityFacade.hasOrganizationRole(org, roleManage)).isTrue();
 
+      List<Role> roles = securityFacade.getOrganizationRoles(org);
+      roles.forEach((item) -> {
+         assertThat(LumeerConst.Security.RESOURCE_ROLES.get(LumeerConst.Security.ORGANIZATION_RESOURCE).contains(item.getName())).isTrue();
+         if (item.getName().equals(roleManage)) {
+            assertThat(item.getUsers().contains(user)).isTrue();
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/controller/SecurityFacadeIntegrationTest.java]: This can be made much simpler:
```java
assertThat(roles).extracting(role -> role.getName())
 .containsOnly(LumeerConst.Security.RESOURCE_ROLES.get(LumeerConst.Security.ORGANIZATION_RESOURCE));

assertThat(roles).filteredOn(role -> role.getName().equals(roleManage))
 .extracting(role -> role.getUsers().contains(user);
```

The same applies to other such blocks.

File: war/src/test/java/io/lumeer/engine/rest/LinkingServiceIntegrationTest.java [hunk 7]
```diff
@@ -89,7 +90,7 @@
    private UserFacade userFacade;
 
    @Before
-   public void init() {
+   public void init() throws InvalidDocumentKeyException {
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/rest/LinkingServiceIntegrationTest.java]: Why is this needed now and was not needed before?

File: war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,1413 @@
...
+            .path(createUrl(
+                  LumeerConst.Security.ORGANIZATION_RESOURCE,
+                  organizationCode,
+                  LumeerConst.Security.ROLES_KEY,
+                  role))
+            .queryParam(USERS_QP, USER_X, USER_Y)
+            .queryParam(GROUPS_QP, GROUP_A, GROUP_B)
+            .request().build(PUT_METHOD_NAME)
+            .invoke();*/
+      // TODO: Checking throwing exception doesn't work.
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java]: Check HTTP codes as we discussed.

File: war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,1413 @@
...
+            .target(TARGET_URI)
+            .path(createUrl(
+                  LumeerConst.Security.ORGANIZATION_RESOURCE,
+                  organizationCode,
+                  LumeerConst.Security.ROLES_KEY))
+            .request().buildGet()
+            .invoke();
+
+      List<Role> fromService = response.readEntity(List.class);
+      assertThat(fromFacade.size()).isEqualTo(fromService.size());
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java]: Please, use collection assertions. It is easier to see where the problem is when such assertion fails.
```java
assertThat(fromService).hasSize(fromFacade.size());
```

File: war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,1413 @@
...
+      for (Role r : fromFacade) {
+         assertThat(fromService.contains(r));
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java]: Make better use of AssertJ:
```java
assertThat(fromService).containsExactly(fromFacade);
```

File: war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,1413 @@
...
+      List<String> groups = null;
+      for (Role item: fromFacade) {
+         if (item.getName().equals(role)) {
+            users = item.getUsers();
+            groups = item.getGroups();
+         }
+      }
+
+      assertThat(users.contains(USER_X)).isTrue();
+      assertThat(users.contains(USER_Y)).isFalse();
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/rest/SecurityServiceIntegrationTest.java]:
```java
assertThat(users).containsOnly(USER_X);
```

File: war/src/main/java/io/lumeer/engine/rest/SecurityService.java [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+         final @QueryParam("groups") List<String> groups) {
+      if (role == null || !isRoleCorrect(LumeerConst.Security.ORGANIZATION_RESOURCE, role))
+         throw new BadRequestException();
+
+      for (String user: users) {
+         securityFacade.addOrganizationUserRole(organizationCode, user, role);
+      }
+
+      for (String group: groups) {
+         securityFacade.addOrganizationGroupRole(organizationCode, group, role);
```
Comment (livthomas) [inline war/src/main/java/io/lumeer/engine/rest/SecurityService.java]: This is highly inefficient since for every user and group we make a new database call. Consider adding appropriate method to the facade to handle such operations.

File: war/src/main/java/io/lumeer/engine/rest/SecurityService.java [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+    * Get roles of an organization.
+    *
+    * @return
+    *       the role object with all roles for the organization
+    * @throws DataFormatException
+    *       if obtained data cannot be transformed to a role object
+    */
+   @GET
+   @Path("/roles")
+   @Produces(MediaType.APPLICATION_JSON)
```
Comment (livthomas) [inline war/src/main/java/io/lumeer/engine/rest/SecurityService.java]: Move all `@Produces` and `@Consumes` annotations to the class level.

File: war/src/main/java/io/lumeer/engine/controller/OrganizationFacade.java (hunk 2 above)
Comment (martinglova) [inline war/src/main/java/io/lumeer/engine/controller/OrganizationFacade.java]: Will do for the next commit.

File: war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java (hunk 3 above)
Comment (martinglova) [inline war/src/main/java/io/lumeer/engine/controller/ProjectFacade.java]: Will do for the next commit.

File: war/src/main/java/io/lumeer/engine/controller/ViewFacade.java (hunk 4 above)
Comment (martinglova) [inline war/src/main/java/io/lumeer/engine/controller/ViewFacade.java]: Will do for the next commit.

File: war/src/main/java/io/lumeer/engine/rest/SecurityService.java [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+    *       the project code to be added the role for
+    * @param role
+    *       the role to be added
+    * @param users
+    *       the users to be added the role for
+    * @param groups
+    *       the groups to be added the role for
+    */
+   @PUT
+   @Path("/projects/{project}/roles/{role}")
```
Comment (livthomas) [inline war/src/main/java/io/lumeer/engine/rest/SecurityService.java]: You should call `projectFacade.setProjectCode(projectCode)` after the initial check in this method as well as all other methods with `project` parameter.

File: war/src/test/java/io/lumeer/engine/controller/SecurityFacadeIntegrationTest.java [hunk 15]
```diff
@@ -106,8 +108,27 @@ public void addAndRemoveOrganizationUserRole() throws Exception {
...
       assertThat(securityFacade.hasOrganizationRole(org, roleManage)).isTrue();
 
+      List<Role> roles = securityFacade.getOrganizationRoles(org);
+      assertThat(roles).extracting(role -> role.getName())
+                       .containsOnly(LumeerConst.Security.RESOURCE_ROLES.get(
+                             LumeerConst.Security.ORGANIZATION_RESOURCE).toArray(new String[LumeerConst.Security.RESOURCE_ROLES.get(
+                                   LumeerConst.Security.ORGANIZATION_RESOURCE).size()]));
+
+      assertThat(roles).filteredOn(role -> role.getName().equals(roleManage))
+                       .extracting(role -> role.getUsers().contains(user));
```
Comment (livthomas) [inline war/src/test/java/io/lumeer/engine/controller/SecurityFacadeIntegrationTest.java]: `extracting()` method itself does not check anything. You need to call `extracting(role -> role.getUsers()).contains(user)`.

File: war/src/main/java/io/lumeer/engine/rest/SecurityService.java (hunk 14 above)
Comment (kubedo8) [inline war/src/main/java/io/lumeer/engine/rest/SecurityService.java]: @livthomas I think that it's not necessary in this case, isn't it?

Comment (livthomas) [inline war/src/main/java/io/lumeer/engine/rest/SecurityService.java]: @kubedo8 probably not but we should rather set it since we rely on this in many parts of our application. I plan to change this whole mechanism around organization and project context in the future.
